File: app/src/main/java/com/trimble/ag/splice/geonote/Database/GeoNoteRepository.java
package com.trimble.ag.splice.geonote.Database;

import android.app.Application;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.room.Room;

import com.trimble.ag.splice.geonote.GeoNote;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import kotlin.jvm.Synchronized;

public class GeoNoteRepository{
    private static final String TAG = "GeoNoteRepo";
    private GeoNoteDao mGeoNoteDao;
    private GeoNoteRepository(GeoNoteDao geoNoteDao){
        mGeoNoteDao = geoNoteDao;
    }
    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private static GeoNoteRepository INSTANCE = null;
    @NonNull
    @Synchronized
    public static GeoNoteRepository getInstance(Context context){
            GeoNoteRepository instance =INSTANCE;
            if(instance==null){
                Log.i(TAG, "Null Repo instance encountered");
                GeoNoteDatabase database = GeoNoteDatabase.getInstance(context);
                instance = new GeoNoteRepository(database.GeoNoteDao());
                INSTANCE =instance;
                Log.i(TAG, "Reset Repo INSTANCE");
            }
            return instance;
    }
    public LiveData<List<GeoNote>> getGeoNotes(){
        return (mGeoNoteDao.getNotes());
    }
    public LiveData<GeoNote> getGeoNote(UUID id) {
        return (mGeoNoteDao.getNote(id));

    }
    public void updateGeoNote(GeoNote geoNote){
        executor.execute(()->{
            mGeoNoteDao.updateGeoNote(geoNote);
        });
    }
    public void deleteGeoNote(GeoNote geoNote){
        executor.execute(()->{
            mGeoNoteDao.deleteGeoNote(geoNote);
        });
    }
    public void addGeoNote(GeoNote geoNote){
        executor.execute(()->{
            mGeoNoteDao.addGeoNote(geoNote);
        });
    }
    /*private GeoNoteDao geoNoteDao;
    private ExecutorService executor = Executors.newSingleThreadExecutor();



    public LiveData<List<GeoNote>> getGeoNotes(){
        return (geoNoteDao.getNotes());
    }
    public LiveData<GeoNote> getGeoNote(UUID id) {
        return (geoNoteDao.getNote(id));

    }
    public void updateGeoNote(GeoNote geoNote){
        executor.execute(()->{
            geoNoteDao.updateGeoNote(geoNote);
        });
    }
    public void deleteGeoNote(GeoNote geoNote){
        executor.execute(()->{
            geoNoteDao.deleteGeoNote(geoNote);
        });
    }
    public void addGeoNote(GeoNote geoNote){
        executor.execute(()->{
            geoNoteDao.addGeoNote(geoNote);
        });
    }*/



}

File: app/src/main/java/com/trimble/ag/splice/geonote/Database/GeoNoteRepository.kt
package com.trimble.ag.splice.geonote.Database

import android.content.Context
import androidx.lifecycle.LiveData
import com.trimble.ag.splice.geonote.GeoNote
import java.io.File
import java.util.*
import java.util.concurrent.Executors

class GeoNoteRepository private constructor (private val geoNoteDao: GeoNoteDao, context: Context)//Used to access and manage the data base
{
    private val executor = Executors.newSingleThreadExecutor()
    private val filesDir = context.applicationContext.filesDir
    //Provides functions that access or query the GeoNote DAO
    fun getGeoNotes(): LiveData<List<GeoNote>> =geoNoteDao.getNotes()
    fun getGeoNote(id: UUID): LiveData<GeoNote?> = geoNoteDao.getNote(id)
    fun updateGeoNote(geoNote: GeoNote){
        executor.execute{
            geoNoteDao.updateGeoNote(geoNote)
        }
    }
    fun addGeoNote(geoNote: GeoNote){
        executor.execute{
            geoNoteDao.addGeoNote(geoNote)
        }
    }
    fun deleteGeoNote(geoNote: GeoNote){
        executor.execute{
            geoNoteDao.deleteGeoNote(geoNote)
        }
    }
    fun getPhotoFile(geoNote: GeoNote):File = File(filesDir, geoNote.photoFileName)
    companion object {//Returns the instance for use in the View Models
        private var INSTANCE: GeoNoteRepository? = null
        fun getInstance(context: Context): GeoNoteRepository {
            synchronized(this) {
                var instance = INSTANCE
                if(instance == null) {
                    val database = GeoNoteDatabase.getInstance(context)
                    instance = GeoNoteRepository(database.geoNoteDao, context)
                    INSTANCE = instance
                }
                return instance
            }
        }
    }
}
